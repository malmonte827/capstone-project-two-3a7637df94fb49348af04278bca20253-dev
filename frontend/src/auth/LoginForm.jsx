import React, { useState } from "react";
import { useNavigate } from "react-router-dom";
import Alert from "../common/Alert"

/** Login Form
 *
 * Shows form and handles update sto state on changes
 *
 * onSubmit calls login function prop
 *
 * Routes -> LoginForm -> Alert
 * Routed as /login
 */

function LoginForm({ login }) {
    const [formData, setFormData] = useState({
        username: "",
        password: "",
    });

    const [formErrors, setFormErrors] = useState([]);

    const navigate = useNavigate()

    /** Handles updating form fields on change */
    function handleChange(evt) {
        const { name, value } = evt.target;
        setFormData((data) => ({ ...data, [name]: value }));
    }

    /** handles form submition
     *
     * If succesful calls signup prop function and if not sets error
     */
    async function handleSubmit(evt) {
        evt.preventDefault();
        try {
            await login(formData);
            navigate("/");
        } catch (err) {
            setFormErrors(err);
        }
    }

    return (
        <div className="LoginForm">
            <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
                <h3 className="mb-3">Login</h3>
                <div className="card">
                    <div className="card-body">
                        <form onSubmit={handleSubmit}>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="username"
                                >
                                    Username
                                </label>
                                <input
                                    className="form-control"
                                    type="text"
                                    name="username"
                                    value={formData.username}
                                    onChange={handleChange}
                                />
                            </div>
                            <div className="mb-3">
                                <label
                                    className="form-label"
                                    htmlFor="password"
                                >
                                    Password
                                </label>
                                <input
                                    className="form-control"
                                    type="password"
                                    name="password"
                                    value={formData.password}
                                    onChange={handleChange}
                                />
                            </div>

                            {formErrors.length ? (
                                <Alert type="danger" messages={formErrors} />
                            ) : null}
                            <div className="d-grid">
                                <button className="btn btn-primary">
                                    Submit
                                </button>
                            </div>
                        </form>
                    </div>
                </div>
            </div>
        </div>
    );
}

export default LoginForm;
