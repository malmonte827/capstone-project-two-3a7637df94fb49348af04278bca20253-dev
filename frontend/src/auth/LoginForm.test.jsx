import React from "react";
import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import "@testing-library/jest-dom";
import { expect, vi } from "vitest";
import LoginForm from "./LoginForm";

vi.mock("../common/Alert");

describe("<LoginForm />", () => {
    const mockLogin = vi.fn();

    beforeEach(() => {
        mockLogin.mockReset();
    });

    it("renders without crashing", () => {
        render(
            <MemoryRouter>
                <LoginForm login={mockLogin} />
            </MemoryRouter>
        );
    });

    it("matches snapshot", () => {
        const { asFragment } = render(
            <MemoryRouter>
                <LoginForm login={mockLogin} />
            </MemoryRouter>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it("handles form input changes", () => {
        render(
            <MemoryRouter>
                <LoginForm login={mockLogin} />
            </MemoryRouter>
        );

        const usernameInput = screen.getByLabelText("Username");
        const passwordInput = screen.getByLabelText("Password");

        fireEvent.change(usernameInput, { target: { value: "testuser" } });
        fireEvent.change(passwordInput, { target: { value: "password123" } });

        expect(usernameInput.value).toBe("testuser");
        expect(passwordInput.value).toBe("password123");
    });

    it("submits the form and calls login function", async () => {
        mockLogin.mockResolvedValueOnce();
        render(
            <MemoryRouter>
                <LoginForm login={mockLogin} />
            </MemoryRouter>
        );

        fireEvent.change(screen.getByLabelText("Username"), {
            target: { value: "testuser" },
        });
        fireEvent.change(screen.getByLabelText("Password"), {
            target: { value: "password123" },
        });

        fireEvent.click(screen.getByText("Submit"));

        await waitFor(() => {
            expect(mockLogin).toHaveBeenCalledWith({
                username: "testuser",
                password: "password123",
            });
        });
    });
});
