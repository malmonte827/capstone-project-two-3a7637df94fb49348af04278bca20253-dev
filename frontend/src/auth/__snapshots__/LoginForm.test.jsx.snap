// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<LoginForm /> > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="LoginForm"
  >
    <div
      class="container col-md-6 offset-md-3 col-lg-4 offset-lg-4"
    >
      <h3
        class="mb-3"
      >
        Login
      </h3>
      <div
        class="card"
      >
        <div
          class="card-body"
        >
          <form>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                autocomplete="username"
                class="form-control"
                id="username"
                name="username"
                required=""
                type="text"
                value=""
              />
            </div>
            <div
              class="mb-3"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                autocomplete="current-password"
                class="form-control"
                id="password"
                name="password"
                required=""
                type="password"
                value=""
              />
            </div>
            <div
              class="d-grid"
            >
              <button
                class="btn btn-primary"
              >
                Submit
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
